src/err/mod.rs

```diff
@@ -193,33 +206,35 @@ impl PyErr {
     /// # Examples
     /// ```rust
     /// use pyo3::prelude::*;
+    /// use pyo3::PyTypeInfo;
     /// use pyo3::exceptions::PyTypeError;
-    /// use pyo3::types::{PyType, PyString};
+    /// use pyo3::types::PyString;
     ///
     /// Python::with_gil(|py| {
     ///     // Case #1: Exception object
-    ///     let err = PyErr::from_value(PyTypeError::new_err("some type error").value(py));
+    ///     let err = PyErr::from_value_bound(PyTypeError::new_err("some type error")
+    ///         .value_bound(py).into_any());
     ///     assert_eq!(err.to_string(), "TypeError: some type error");
     ///
     ///     // Case #2: Exception type
-    ///     let err = PyErr::from_value(PyType::new::<PyTypeError>(py));
+    ///     let err = PyErr::from_value_bound(PyTypeError::type_object_bound(py).into_any());
     ///     assert_eq!(err.to_string(), "TypeError: ");
     ///
     ///     // Case #3: Invalid exception value
-    ///     let err = PyErr::from_value(PyString::new_bound(py, "foo").as_gil_ref());
+    ///     let err = PyErr::from_value_bound(PyString::new_bound(py, "foo").into_any());
     ///     assert_eq!(
     ///         err.to_string(),
     ///         "TypeError: exceptions must derive from BaseException"
     ///     );
     /// });
     /// ```
-    pub fn from_value(obj: &PyAny) -> PyErr {
+    pub fn from_value_bound(obj: Bound<'_, PyAny>) -> PyErr {
         let state = if let Ok(obj) = obj.downcast::<PyBaseException>() {
-            PyErrState::normalized(obj)
+            PyErrState::normalized(obj.clone())
         } else {
             // Assume obj is Type[Exception]; let later normalization handle if this
             // is not the case
-            PyErrState::lazy(obj, obj.py().None())
+            PyErrState::lazy_bound(obj.clone(), obj.py().None())
```

Comment:
I think the clones should not be necessary in here, probably by using `.downcast_into` and by calling `ob.py().None()` before constructing the state.

Reply:
I couldn't get `.downcast_into` to work - that takes ownership of `obj`, so it's unavailable for the `else` branch.

Reply:
Ah I see; let's just get #3829 merged and then we should be able to use that here!

Reply:
I tried doing:

```rust
    pub fn from_value_bound(obj: &Bound<'_, PyAny>) -> PyErr {
        let state = match obj.downcast_into::<PyBaseException>() {
            Ok(obj) => PyErrState::normalized(obj),
            Err(err) => {
                // Assume obj is Type[Exception]; let later normalization handle if this
                // is not the case
                let obj = err.into_inner();
                let none = obj.py().None();
                PyErrState::lazy_bound(&obj, none)
            }
        };

        PyErr::from_state(state)
    }
```

But that gets me this compiler error:

```
error[E0507]: cannot move out of `*obj` which is behind a shared reference
    --> src/err/mod.rs:239:27
     |
239  |         let state = match obj.downcast_into::<PyBaseException>() {
     |                           ^^^ ---------------------------------- `*obj` moved due to this method call
     |                           |
     |                           move occurs because `*obj` has type `instance::Bound<'_, types::any::PyAny>`, which does not implement the `Copy` trait
     |
note: `types::any::PyAnyMethods::downcast_into` takes ownership of the receiver `self`, which moves `*obj`
    --> src/types/any.rs:1585:25
     |
1585 |     fn downcast_into<T>(self) -> Result<Bound<'py, T>, DowncastIntoError<'py>>
     |                         ^^^^
help: you can `clone` the value and consume it, but this might not be your desired behavior
     |
239  |         let state = match obj.clone().downcast_into::<PyBaseException>() {
     |                              ++++++++
```

Reply:
If I add the suggested `.clone()` the compiler and local tests are happy again, but since we're trying to avoid `.clone()` and the compiler says cloning might not be desired I haven't pushed that change.

Reply:
After implementing the Bound constructors for `PyUnicodeDecodeError` I hit this error:

```rust
error[E0308]: mismatched types
   --> src/exceptions.rs:27:49
    |
27  |                   $crate::PyErr::from_value_bound(err)
    |                   ------------------------------- ^^^ expected `Bound<'_, PyAny>`, found `&CustomError`
    |                   |
    |                   arguments to this function are incorrect
...
974 | /         create_exception!(
975 | |             mymodule,
976 | |             CustomError,
977 | |             PyException,
978 | |             concat!("Some", " more ", stringify!(docs))
979 | |         );
    | |_________- in this macro invocation
    |
    = note: expected struct `instance::Bound<'_, types::any::PyAny>`
            found reference `&exceptions::tests::custom_exception_doc_expr::CustomError`
note: associated function defined here
   --> src/err/mod.rs:239:12
    |
239 |     pub fn from_value_bound(obj: Bound<'_, PyAny>) -> PyErr {
    |            ^^^^^^^^^^^^^^^^ ---------------------
    = note: this error originates in the macro `$crate::impl_exception_boilerplate` which comes from the expansion of the macro `create_exception` (in Nightly builds, run with -Z macro-backtrace for more info)
```

And at this point I'm not clear what the way forward is.

Reply:
That's a `From` implementation converting an GIL Ref to an exception instance into a `PyErr`, defined in the macro `impl_exception_boilerplate`.

Probably you should leave that one on line 27 as `from_value` with an `#[allow(deprecated)]`.

Probably it's worth also adding `From<Bound<'_, $name> for PyErr` to that macro too.

Reply:
`#[allow(deprecated)]` did the trick!

I'm running into trouble with the `From<Bound<'_, $name>> for PyErr` though. I tried adding:

```rust
        impl<'py> ::std::convert::From<$crate::Bound<'py, $name>> for $crate::PyErr {
            #[inline]
            fn from(err: $crate::Bound<'py, $name>) -> $crate::PyErr {
                $crate::PyErr::from_value_bound(err.into_any())
            }
        }
```

But I end up hitting orphan rules:

```rust
---- src/lib.rs - doc_test::guide_exception_md (line 579) stdout ----
error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
  --> src/lib.rs:588:5
   |
13 |     pyo3::import_exception!(io, UnsupportedOperation);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     impl doesn't use only types from inside the current crate
   |     `PyErr` is not defined in the current crate
   |     `pyo3::Bound` is not defined in the current crate
   |
   = note: define and implement a trait or new type instead
   = note: this error originates in the macro `$crate::impl_exception_boilerplate` which comes from the expansion of the macro `pyo3::import_exception` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I'm not clear what the right way to fix this is.

Reply:
Oof, of course, let's drop the `From<Bound<$name>> for PyErr` for now then and defer to a follow-up PR.

Probably the way to resolve it would be to add a generic implementation in PyO3 bounded on some trait which these downstream-defined types can implement. Would need to play around.

Reply:
I think this is ready then!